### PR TITLE
Allow CID to fallback on cookie creation if explicitly allowed by the use case

### PR DESCRIPTION
**Pull request by dvoytenko**
Implements #1382.

### CODE REVIEW

File: extensions/amp-access/0.1/amp-access.js [hunk 1]
```diff
@@ -66,6 +66,9 @@ const VIEW_TIMEOUT = 2000;
 /** @const {string} */
 const TEMPLATE_PROP = '__AMP_ACCESS__TEMPLATE';
 
+/** @const {time} */
+const COOKIE_EXPIRATION_TIME = 365 * 24 * 60 * 60 * 1000;  // 1 year.
```
Comment (cramforce) [inline extensions/amp-access/0.1/amp-access.js]: Can we share this with localStorage impl.?

File: src/service/cid-impl.js [hunk 2]
```diff
@@ -78,19 +86,24 @@ class Cid {
...
+   *     this parameter enables CID to create a cookie if it doesn't yet
+   *     exist.
+   * @return {!Promise<?string>} A client identifier that should be used
    *      within the current source origin and externalCidScope. Might be
    *      null if no identifier was found or could be made.
    *      This promise may take a long time to resolve if consent isn't
    *      given.
    */
-  get(externalCidScope, consent, opt_persistenceConsent) {
+  get(externalCidScope, consent, opt_persistenceConsent,
```
Comment (cramforce) [inline src/service/cid-impl.js]: I was thinking to maybe convert the first arg to a prop object. `opt_persistenceConsent` will be undefined for most callers (all but 1) while opt_createCookieParams should be set quite often

Comment (cramforce) [inline src/service/cid-impl.js]: We can support both string and object to reduce patch size, if that helps.

File: src/service/cid-impl.js [hunk 3]
```diff
@@ -118,6 +136,42 @@ function getExternalCid(cid, externalCidScope, persistenceConsent) {
...
+ */
+function getOrCreateCookie(cid, externalCidScope,
+    persistenceConsent, createCookieParams) {
+  const win = cid.win;
+  const existingCookie = getCookie(win, externalCidScope);
+  if (existingCookie || !createCookieParams) {
+    return Promise.resolve(existingCookie);
+  }
+
+  // Create new cookie, always prefixed with "amp-".
```
Comment (cramforce) [inline src/service/cid-impl.js]: , so that we can see from the value whether we created it.

File: src/service/cid-impl.js (hunk 2 above)
Comment (dvoytenko) [inline src/service/cid-impl.js]: I see. I thought you hated those :)

**Comment by cramforce** [comment]
Looks fine to me.

File: extensions/amp-access/0.1/amp-access.js (hunk 1 above)
Comment (dvoytenko) [inline extensions/amp-access/0.1/amp-access.js]: I can definitely use `BASE_CID_MAX_AGE_MILLIS`.

Comment (dvoytenko) [inline extensions/amp-access/0.1/amp-access.js]: Done

File: src/service/cid-impl.js (hunk 2 above)
Comment (dvoytenko) [inline src/service/cid-impl.js]: Done

File: src/service/cid-impl.js (hunk 3 above)
Comment (dvoytenko) [inline src/service/cid-impl.js]: Done

**Comment by dvoytenko** [comment]
@cramforce changes made and tests added. PTAL.

File: src/service/cid-impl.js [hunk 4]
```diff
@@ -100,21 +120,54 @@ class Cid {
...
+function getOrCreateCookie(cid, getCidStruct, persistenceConsent) {
+  const win = cid.win;
+  const existingCookie = getCookie(win, getCidStruct.scope);
+  if (existingCookie || !getCidStruct.createCookieIfNotPresent) {
+    return Promise.resolve(existingCookie);
+  }
+
+  // Create new cookie, always prefixed with "amp-", so that we can see from
+  // the value whether we created it.
+  const newCookie = 'amp-' + cid.sha384Base64_(getEntropy(win));
```
Comment (jridgewell) [inline src/service/cid-impl.js]: Is it necessary to have entropy on this cookie?

Comment (dvoytenko) [inline src/service/cid-impl.js]: I would think so - we still want a cryptographically strong and unique value, similar to what's produced for CID in other scenarios. @cramforce WDYT?

Comment (dvoytenko) [inline src/service/cid-impl.js]: Basically `getEntropy` is `getRandomValue`...

Comment (cramforce) [inline src/service/cid-impl.js]: @jridgewell The value should be to be extremely likely to be unique (and hard to guess another one). Little reason to use a different strategy from the one we use for localStorage.

Comment (jridgewell) [inline src/service/cid-impl.js]: My thinking is that the one we use in local storage has to be unique so separate websites cannot determine this is the same user. Here we just need a sentinel value, right?

**Comment by cramforce** [comment]
LGTM

I updated the description to not close the bug, because we still need to expose this to analytics.

File: src/service/cid-impl.js (hunk 4 above)
Comment (dvoytenko) [inline src/service/cid-impl.js]: No - the meaning doesn't change because we change the storage mechanism.

**Pull request merged**